warehouse/macaroons/caveats.py

```diff
@@ -145,9 +146,30 @@ def verify(self, key):
         self.verifier.satisfy_general(ProjectIDsCaveat(self))
 
         try:
-            return self.verifier.verify(self.macaroon, key)
+            result = self.verifier.verify(self.macaroon, key)
         except (
             pymacaroons.exceptions.MacaroonInvalidSignatureException,
             Exception,  # https://github.com/ecordell/pymacaroons/issues/50
-        ):
-            return False
+        ) as exc:
+            failure_reasons = []
+            for cb in self.verifier.callbacks:
+                failure_reason = getattr(cb, "failure_reason", None)
+                if failure_reason is not None:
+                    failure_reasons.append(failure_reason)
+
+            # If we have a more detailed set of failure reasons, use them.
+            # Otherwise, use whatever the exception gives us.
+            if len(failure_reasons) > 0:
+                return WarehouseDenied(
+                    ", ".join(failure_reasons), reason="invalid_api_token"
+                )
+            else:
+                return WarehouseDenied(str(exc), reason="invalid_api_token")
+
+        # NOTE: We should never hit this case, since pymacaroons *should* always either
+        # raise on failure *or* return true. But there's nothing stopping that from
+        # silently breaking in the future, so we check the result defensively here.
+        if not result:
+            return WarehouseDenied("unknown error", reason="invalid_api_token")
```

Comment:
Flagging for review: this is a consequence of pymacaroons' API design: they return `True` if verification succeeds, and raise on any verification error. As a result, they could *conceivably* return `False` in a future version and, without this, we'd incorrectly accept an invalid token.

At the moment, however, it's a no-op.

Reply:
@ewjoachim identified this same problem in an upstream issue: https://github.com/ecordell/pymacaroons/issues/55